File: ATZAlertController/ATZAlertController.m
// ATZAlertController.m
//
// Copyright (c) 2016 Shashank Pali
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

#import "ATZAlertController+Helper.h"

@implementation ATZAlertController

- (void)show
{
  [self showAlertAnimation:YES];
}

- (void)showWithAnimation:(BOOL)animated
{
  [self showAlertAnimation:animated];
}

#pragma mark - View Disapper

- (void)viewDidDisappear:(BOOL)animated
{
  [super viewDidDisappear:animated];
  
  // precaution to insure window gets destroyed
  self.alertWindow.hidden = YES;
  self.alertWindow = nil;
}

#pragma mark - instance Method

+ (nonnull instancetype)alertWithTitle:(NSString *)title
                               message:(NSString *)message
                     cancelButtonTitle:(NSString *)cancel
                destructiveButtonTitle:(NSString *)destructive
                     otherButtonTitles:(NSArray<NSString *> *)titleArray
                        preferredStyle:(UIAlertControllerStyle)style
                    showAlertInstantly:(BOOL)showAlert
                                 block:(actionHandler)handler
{
  for (id obj in titleArray) {
    
    NSAssert([obj isKindOfClass:[NSString class]], @"Required array containing \"String\" as object");
  }
  
  ATZAlertController *alert = [ATZAlertController title:title message:message cancelButtonTitle:cancel destructiveButtonTitle:destructive otherButtonTitles:titleArray preferredStyle:style block:^(UIAlertAction * _Nonnull action, NSString * _Nonnull titleString)
                               {
                                 if (handler) {
                                   handler(action , titleString);
                                 }
                               }];
  if (showAlert)
  {
    [alert show];
  }
  return alert;
}

+ (nonnull instancetype)alertWithTitle:(NSString * _Nonnull)title
                               message:(NSString * _Nonnull)message
                     cancelButtonTitle:(NSString * _Nonnull)cancel
                     otherActionButton:(NSArray<ATZActionButton *> * _Nonnull)buttonArray
                        preferredStyle:(UIAlertControllerStyle)style
                    showAlertInstantly:(BOOL)showAlert
                                 block:(nullable actionHandler)handler
{
  for (id obj in buttonArray) {
    
    NSAssert([obj isKindOfClass:[ATZActionButton class]], @"Required array containing \"ATZActionButton\" as object");
  }
  
  ATZAlertController *alert =[ATZAlertController title:title message:message cancelButtonTitle:cancel destructiveButtonTitle:nil otherButtonTitles:buttonArray preferredStyle:style block:^(UIAlertAction * _Nonnull action, NSString * _Nonnull titleString)
                              {
                                if (handler) {
                                  handler(action , titleString);
                                }
                              }];
  if (showAlert)
  {
    [alert show];
  }
  return alert;
}

+ (nonnull instancetype)alertWithTitle:(NSString * _Nonnull)title
                               message:(NSString * _Nonnull)message
                     cancelButtonTitle:(NSString * _Nonnull)cancel
                    showAlertInstantly:(BOOL)showAlert
{
  ATZAlertController *alert =[ATZAlertController title:title message:message cancelButtonTitle:cancel destructiveButtonTitle:nil otherButtonTitles:nil preferredStyle:UIAlertControllerStyleAlert block:nil];
  if (showAlert)
  {
    [alert show];
  }
  return alert;
}

@end

#pragma mark - ATZActionButton

@implementation ATZActionButton

+ (ATZActionButton * _Nonnull)actionWithTitle:(NSString * _Nonnull)title preferredStyle:(UIAlertActionStyle)style{
  
  ATZActionButton *alertButton = [[ATZActionButton alloc] init];
  alertButton.title = title;
  alertButton.style = style;
  
  return alertButton;
}

@end

File: ATZAlertController->Swift/ATZAlertController.swift
// ATZAlertController.swift
//
// Copyright (c) 2016 Shashank Pali
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.
//

import UIKit
import Foundation

public typealias actionHandler = (action: UIAlertAction, titleString: String) -> Void

//MARK:- ATZAlertController Class
public class ATZAlertController: UIAlertController
{
  override public func viewDidDisappear(animated: Bool)
  {
    alertWindow.hidden = true
    alertWindow = nil
  }
  /**
   Call show alert
   */
  public func show()
  {
    showAlertAnimation(true)
  }
  /**
   Call show alert with animation option
   
   - parameter animated: Derive whether to animate or not.
   */
  public func showWithAnimation(animated: Bool)
  {
    showAlertAnimation(animated)
  }
  /**
   Show alert with string array
   
   - parameter title:                  Alert title.
   - parameter message:                Alert message.
   - parameter cancelButtonTitle:      Cancel title (Tapping leads to dismiss controller ONLY).
   - parameter destructiveButtonTitle: Destructive title (Tapping leads to CALLBACK in block).
   - parameter otherButtonTitles:      Create other button with respect to array (Tapping leads to CALLBACK
   in block respectively).
   - parameter preferredStyle:         Style description(Declare whether its Alert or ActionSheet).
   - parameter showAlertInstantly:     Derive to show alert immediately or leave it to user to call SHOW
   - parameter handler:                Handler description(Return the respective action and title of tapped button).
   
   
   - returns: return ATZAlertController object
   */
  public class func alertWithTitle(title: String,
                                   message: String,
                                   cancelButtonTitle: String,
                                   destructiveButtonTitle: String?,
                                   otherButtonTitles: [String]?,
                                   preferredStyle: UIAlertControllerStyle,
                                   showAlertInstantly: Bool,
                                   handler: actionHandler?) -> ATZAlertController
  {
    let alert = ATZAlertController.alertController(title,
                                                   message: message,
                                                   cancel: cancelButtonTitle,
                                                   destructive: destructiveButtonTitle,
                                                   titleArray: otherButtonTitles,
                                                   style: preferredStyle) {
                                                    
                                                    handler?(action: $0, titleString: $1)
                                                    
    }
    if (showAlertInstantly)
    {
      alert.show()
    }
    return alert
  }
  /**
   Show alert with ATZActionButton array
   
   - parameter title:              Alert title.
   - parameter message:            Alert message.
   - parameter cancelButtonTitle:  Cancel title (Tapping leads to dismiss controller ONLY).
   - parameter otherActionButton:  Create other button with respect to array (Tapping leads to CALLBACK
   in block respectively).
   - parameter preferredStyle:     Style description(Declare whether its Alert or ActionSheet).
   - parameter showAlertInstantly: Derive to show alert immediately or leave it to user to call SHOW
   - parameter handler:            Handler description(Return the respective action and title of tapped button).
   
   - returns: return ATZAlertController object
   */
  public class func alertWithTitle(title: String,
                                   message: String,
                                   cancelButtonTitle: String,
                                   otherActionButton: [ATZActionButton],
                                   preferredStyle: UIAlertControllerStyle,
                                   showAlertInstantly: Bool,
                                   handler: actionHandler?) -> ATZAlertController
  {
    let alert = ATZAlertController.alertController(title,
                                                   message: message,
                                                   cancel: cancelButtonTitle,
                                                   destructive: nil,
                                                   titleArray: otherActionButton,
                                                   style: preferredStyle) {
                                                    
                                                    handler?(action: $0, titleString: $1)
                                                    
    }
    if (showAlertInstantly)
    {
      alert.show()
    }
    return alert
  }
  /**
   Show simple alert
   
   - parameter title:              Alert title.
   - parameter message:            Alert message.
   - parameter cancelButtonTitle:  Cancel title (Tapping leads to dismiss controller ONLY).
   - parameter showAlertInstantly: Derive to show alert immediately or leave it to user to call SHOW
   - parameter handler:            Handler description(Return the respective action and title of tapped button).
   
   - returns: return ATZAlertController object
   */
  public class func alertWithTitle(title: String,
                                   message: String,
                                   cancelButtonTitle: String,
                                   showAlertInstantly: Bool,
                                   handler: actionHandler?) -> ATZAlertController
  {
    let alert = ATZAlertController.alertController(title,
                                                   message: message,
                                                   cancel: cancelButtonTitle,
                                                   destructive: nil,
                                                   titleArray: nil,
                                                   style: .Alert) {
                                                    
                                                    handler?(action: $0, titleString: $1)
                                                    
    }
    if (showAlertInstantly)
    {
      alert.show()
    }
    return alert
  }
  /**
   Show Login Alert
   
   - parameter title:                    Alert title.
   - parameter message:                  Alert message.
   - parameter userNamePlaceholderTitle: Shows placeholder text.
   - parameter passwordPlaceholderTitle: Shows placeholder text.
   - parameter showAlertInstantly:       Derive to show alert immediately or leave it to user to call SHOW
   - parameter handler:                  Handler description(Return the respective action and title of tapped button).
   
   - returns: return ATZAlertController object
   */
  public class func loginAlertWithTitle(title: String,
                                        message: String,
                                        userNamePlaceholderTitle: String,
                                        passwordPlaceholderTitle: String,
                                        showAlertInstantly: Bool,
                                        handler: actionHandler?) -> ATZAlertController
  {
    let alert = ATZAlertController.alertController(title,
                                                   message: message,
                                                   cancel: nil,
                                                   destructive: nil,
                                                   titleArray: ["Login","Cancel"],
                                                   style: .Alert) {
                                                    handler?(action: $0, titleString: $1)
    }
    alert.addTextFieldWithConfigurationHandler {
      $0.placeholder = userNamePlaceholderTitle
    }
    alert.addTextFieldWithConfigurationHandler {
      $0.placeholder = passwordPlaceholderTitle
      $0.secureTextEntry = true
    }
    if (showAlertInstantly)
    {
      alert.show()
    }
    return alert
  }
  
  
}

//MARK:- ATZActionButton Class
public class ATZActionButton: NSObject
{
  var title: String!
  var style: UIAlertActionStyle!
  /**
   Create ATZActionButton (Recommended for [Show alert with ATZActionButton array]
   function).
   - parameter title: Button title
   - parameter style: Button style(Derive whether its default,cancel or destructive).
   - returns: return ATZActionButton object.
   */
  public class func actionWithTitleAndPrefferedStyle(title: String, style: UIAlertActionStyle) -> ATZActionButton
  {
    let actionButton = ATZActionButton()
    actionButton.title = title
    actionButton.style = style
    
    return actionButton
  }
}
